### PR TITLE
Test native images in CI

### LINKED ISSUE #13840: Test native images in CI
### Problem

Native images are slow to build so we currently only build for a tag, but that's too late in the lifecycle to catch any bugs specific to the native image.

### Solution

* Add a workflow similar to acceptance but with native image builds similar to release-ghcr
* Ensure it runs nightly and on every tag
* Look at ways to extract common logic to [reusable](https://docs.github.com/en/actions/how-tos/reuse-automations/reuse-workflows) workflows to stay DRY

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by bilyana-gospodinova**
**Description**:
This PR adds a new workflow that will test periodically if the native images are built successfully and that they pass the acceptance tests. The workflow is triggered:
- at 05:00 am UTC
- on every tag

Fixes #13840

### CODE REVIEW

File: .github/workflows/test-native-images.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    inputs:
+      branch:
+        description: "Branch"
+        required: true
+        type: string
+
+permissions:
+  contents: read
+  packages: read
+  pull-requests: read
```
Comment (steven-sheehy) [inline .github/workflows/test-native-images.yml]: Probably doesn't need this permission

File: .github/workflows/release-ghcr.yml [hunk 2]
```diff
@@ -108,23 +88,30 @@ jobs:
...
 
-      - name: Get Version
+      - name: Get version
         run: echo "VERSION=$(grep -oP 'version=\K.+' gradle.properties)" >> $GITHUB_ENV
 
       - name: Login to GHCR
         uses: docker/login-action@af1e73f918a031802d376d3c8bbc3fe56130a9b0 # v4.4.0
         with:
-          password: ${{ secrets.GITHUB_TOKEN }}
+          password: ${{ github.token }}
```
Comment (steven-sheehy) [inline .github/workflows/release-ghcr.yml]: Why this change?

File: .github/actions/release-native-image/action.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        distribution: temurin
+        java-version: 25
+
+    - name: Setup Gradle
+      uses: gradle/actions/setup-gradle@3f131e8634966bd73d06cc69884922b02e6faf92 # v4
+
+    - name: Determine version
+      shell: bash
+      run: |
+        echo "IMAGE_REGISTRY=ghcr.io/hiero-ledger/hiero-mirror-node" >> $GITHUB_ENV
```
Comment (steven-sheehy) [inline .github/actions/release-native-image/action.yml]: unused

File: .github/workflows/release-ghcr.yml [hunk 4]
```diff
@@ -51,29 +49,11 @@ jobs:
       - name: Checkout Code
         uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0 # v7.0.0
 
-      - name: Get version
-        run: echo "VERSION=$(grep -oP 'version=\K.+' gradle.properties)" >> $GITHUB_ENV
-
-      - name: Install JDK
-        uses: actions/setup-java@0f481fcb613427c0f801b606911222b5b6f3083a # v5.5.0
+      - name: Release native image
+        uses: ./.github/actions/release-native-image
```
Comment (steven-sheehy) [inline .github/workflows/release-ghcr.yml]: Why was this made reusable but not used in multiple places? Should revert all changes to release-ghcr and remove release-native-image.

File: .github/actions/build-native-images/action.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        username: ${{ github.actor }}
+
+    - name: Build and push native images
+      if: env.TRIGGERED_BY_TAG != 'true'
+      shell: bash
+      run: |
+        set -ex
+        for module in grpc importer monitor rest-java web3; do
+          ./gradlew ":${module}:bootBuildImage" -PimagePlatform=linux/amd64
+          docker push "${IMAGE_REGISTRY}/${module}:${VERSION}"
```
Comment (steven-sheehy) [inline .github/actions/build-native-images/action.yml]: This pushes actual images to ghcr, which is a change from existing acceptance. Existing one just builds the images into local docker then loads those into kind. I think we should stick with this approach as we don't want to push snapshot images to ghcr or have race conditions with multiple workflows writing to same tag. Should revert login step as well.

File: .github/workflows/test-native-images.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        with:
+          egress-policy: audit
+
+      - name: Checkout
+        uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0 # v7.0.0
+        with:
+          ref: ${{ github.event_name == 'workflow_dispatch' && github.event.inputs.branch || '' }}
+
+      - name: Build native images
+        uses: ./.github/actions/build-native-images
```
Comment (steven-sheehy) [inline .github/workflows/test-native-images.yml]: This action is only used by this workflow, so it doesn't really need to be separate.

File: .github/actions/build-native-images/action.yml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        registry: ghcr.io
+        username: ${{ github.actor }}
+
+    - name: Build and push native images
+      if: env.TRIGGERED_BY_TAG != 'true'
+      shell: bash
+      run: |
+        set -ex
+        for module in grpc importer monitor rest-java web3; do
+          ./gradlew ":${module}:bootBuildImage" -PimagePlatform=linux/amd64
```
Comment (steven-sheehy) [inline .github/actions/build-native-images/action.yml]: This will be slow because it's not done in parallel and each build takes 10+ minutes. Should be done as a matrix step like in release-ghcr.

File: .github/workflows/acceptance.yaml [hunk 8]
```diff
@@ -23,235 +23,24 @@ on:
 
 permissions:
   contents: read
-
-defaults:
-  run:
-    shell: bash
-
-env:
-  LC_ALL: C.UTF-8
```
Comment (steven-sheehy) [inline .github/workflows/acceptance.yaml:32]: Why are we removing these?

File: .github/workflows/acceptance.yaml [hunk 9]
```diff
@@ -23,235 +23,24 @@ on:
...
-          kubectl describe pods -n "${SOLO_NAMESPACE}" || echo "Could not describe pods"
-
-          echo "--- Fetching logs for all pods in namespace ${SOLO_NAMESPACE} ---"
-          for pod in $(kubectl get pods -n "${SOLO_NAMESPACE}" -o name); do
-            if [ -n "$pod" ]; then
-              echo "--- Logs for pod: $pod ---"
-              kubectl logs "$pod" -n "${SOLO_NAMESPACE}" --all-containers=true --tail=10000 || echo "Could not get logs for pod $pod"
-            fi
-          done
+    uses: ./.github/workflows/run-e2e.yml
```
Comment (steven-sheehy) [inline .github/workflows/acceptance.yaml]: Is it possible to use this run-e2e also in wrb-differential?

File: .github/workflows/run-e2e.yml [hunk 10]
```diff
@@ -0,0 +1,327 @@
+# SPDX-License-Identifier: Apache-2.0
+
+name: Run end-to-end tests
+
+on:
+  workflow_call:
```
Comment (steven-sheehy) [inline .github/workflows/run-e2e.yml]: Why is this one a workflow instead of an action?

File: .github/workflows/run-e2e.yml [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+                while ! docker manifest inspect "gcr.io/mirrornode/hedera-mirror-${module}:${TAG}" >/dev/null 2>&1; do
+                  sleep 10
+                done
+              done
+            else
+              for module in grpc importer monitor rest rest-java test web3; do
+                while ! docker manifest inspect "gcr.io/mirrornode/hedera-mirror-${module}:${TAG}" >/dev/null 2>&1; do
+                  sleep 10
+                done
+              done
```
Comment (steven-sheehy) [inline .github/workflows/run-e2e.yml]: Having a separate wait-for-images action would be very useful in other places like helm chart as well. Maybe we can add that and have inputs as native=true|false and version.

File: .github/workflows/run-e2e.yml [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+            fi
+          fi
+
+      - name: Find changed modules
+        id: changeset
+        if: env.TRIGGERED_BY_TAG != 'true' && inputs.build-changed-modules-only == 'true'
+        uses: step-security/paths-filter@5c5241b8233e77b55b9046daf88f1cb7560281de # v4.0.1
+        with:
+          base: ${{ env.BASE }}
```
Comment (steven-sheehy) [inline .github/workflows/run-e2e.yml]: It looks like base is set to either "" or v$TAG but latter is skipped in this step. So do we even need BASE?

File: .github/workflows/run-e2e.yml [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+        if: env.TRIGGERED_BY_TAG != 'true' && (inputs.image-source == 'native' || (inputs.build-changed-modules-only == 'true' && steps.changeset.outputs.changes != '[]'))
+        uses: gradle/actions/setup-gradle@3f131e8634966bd73d06cc69884922b02e6faf92 # v4
+
+      - name: Login to GHCR
+        if: inputs.image-source == 'native'
+        uses: docker/login-action@af1e73f918a031802d376d3c8bbc3fe56130a9b0 # v4.4.0
+        with:
+          password: ${{ secrets.GITHUB_TOKEN }}
+          registry: ghcr.io
+          username: ${{ github.actor }}
```
Comment (steven-sheehy) [inline .github/workflows/run-e2e.yml]: Can be removed

File: .github/workflows/run-e2e.yml [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+  IMAGE_REGISTRY: ghcr.io/hiero-ledger/hiero-mirror-node
+
+jobs:
+  e2e:
+    runs-on: hiero-mirror-node-linux-large
+    timeout-minutes: ${{ inputs.timeout-minutes }}
+    permissions:
+      contents: read
+      packages: read
+      pull-requests: read
```
Comment (steven-sheehy) [inline .github/workflows/run-e2e.yml]: Do we need either?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: .github/workflows/acceptance.yaml (hunk 8 above)
Comment (bilyana-gospodinova) [reply .github/workflows/acceptance.yaml:32]: They were not needed anymore as this workflow was changed to call `run-e2e.yml` and it has these lines defined in it.

File: .github/workflows/run-e2e.yml (hunk 13 above)
Comment (bilyana-gospodinova) [reply .github/workflows/run-e2e.yml]: If we build simultaneously all modules with a matrix as described here - https://github.com/hiero-ledger/hiero-mirror-node/pull/13908#discussion_r3599310875, each module is built in a different runner and in order to access the built images in the acceptance tests, the images need to be pushed and then pulled.

File: .github/workflows/run-e2e.yml (hunk 12 above)
Comment (bilyana-gospodinova) [reply .github/workflows/run-e2e.yml]: It is needed only in the corner case where the acceptance workflow runs, triggered on "release/*" and the HEAD is different from the from the release version top commit.

File: .github/workflows/test-native-images.yml (hunk 1 above)
Comment (bilyana-gospodinova) [reply .github/workflows/test-native-images.yml]: Removed

File: .github/workflows/release-ghcr.yml (hunk 2 above)
Comment (bilyana-gospodinova) [reply .github/workflows/release-ghcr.yml]: Reverted

File: .github/actions/release-native-image/action.yml (hunk 3 above)
Comment (bilyana-gospodinova) [reply .github/actions/release-native-image/action.yml]: Removed

File: .github/workflows/release-ghcr.yml (hunk 4 above)
Comment (bilyana-gospodinova) [reply .github/workflows/release-ghcr.yml]: Addressed

File: .github/workflows/test-native-images.yml (hunk 6 above)
Comment (bilyana-gospodinova) [reply .github/workflows/test-native-images.yml]: Currently it is used in 2 places.

File: .github/actions/build-native-images/action.yml (hunk 7 above)
Comment (bilyana-gospodinova) [reply .github/actions/build-native-images/action.yml]: Added matrix.

File: .github/workflows/acceptance.yaml (hunk 9 above)
Comment (bilyana-gospodinova) [reply .github/workflows/acceptance.yaml]: It would require to have too much if-else logic especially in the solo cluster setup, so maybe we should not do it for now.

File: .github/workflows/run-e2e.yml (hunk 10 above)
Comment (bilyana-gospodinova) [reply .github/workflows/run-e2e.yml]: Changed to action.

File: .github/workflows/run-e2e.yml (hunk 11 above)
Comment (bilyana-gospodinova) [reply .github/workflows/run-e2e.yml]: Created a wait-for-images action.

File: .github/workflows/run-e2e.yml (hunk 12 above)
Comment (bilyana-gospodinova) [reply .github/workflows/run-e2e.yml]: Update: After the latest changes it is not needed anymore. Removed.

File: .github/workflows/run-e2e.yml (hunk 14 above)
Comment (bilyana-gospodinova) [reply .github/workflows/run-e2e.yml]: Removed

File: .github/actions/build-native-images/action.yml (hunk 5 above)
Comment (bilyana-gospodinova) [reply .github/actions/build-native-images/action.yml]: Building the images in parallel uses one type of runner, then running the acceptance tests requires another type of runner. I changed it to upload the native images into a temporary artifact so that the images can be passed between runners.

File: monitor/src/main/java/org/hiero/mirror/monitor/publish/TransactionPublisher.java [hunk 15]
```diff
@@ -36,7 +38,15 @@ public class TransactionPublisher implements AutoCloseable {
...
-    private final Flux<Client> clients = Flux.defer(this::getClients).cache();
+    // Retry before cache so a transient empty node list is not permanently cached as an error.
+    private final Flux<Client> clients = Flux.defer(this::getClients)
+            .retryWhen(Retry.backoff(Long.MAX_VALUE, Duration.ofSeconds(1L))
+                    .maxBackoff(Duration.ofSeconds(8L))
+                    .doBeforeRetry(r -> log.warn(
+                            "Retrying client creation attempt #{} after failure: {}",
+                            r.totalRetries() + 1,
+                            r.failure().getMessage())))
+            .cache();
```
Comment (steven-sheehy) [inline monitor/src/main/java/org/hiero/mirror/monitor/publish/TransactionPublisher.java]: Revert changes to monitor code as I have a different fix in the works via #13937

File: .github/workflows/acceptance.yaml (hunk 8 above)
Comment (steven-sheehy) [reply .github/workflows/acceptance.yaml:32]: But there's still other steps before the call to run-e2e including checkout. We want to ensure the files are checked out with the correct encoding. And it's not clear to me that setting these inside an action will propagate them to steps in the workflow that occur after the action. Please revert

File: .github/actions/run-e2e/action.yml [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+    default: "false"
+    required: false
+  dockerhub-token:
+    description: Docker Hub token for rate-limit auth (optional)
+    default: ""
+    required: false
+  dockerhub-username:
+    description: Docker Hub username for rate-limit auth (optional)
+    default: ""
+    required: false
```
Comment (steven-sheehy) [inline .github/actions/run-e2e/action.yml]: Should not pass these as inputs. See https://docs.github.com/en/actions/how-tos/reuse-automations/reuse-workflows#passing-secrets-to-nested-workflows. Think you can just use `secrets: inherit` and use the secret as normal. Username isn't a secret so I'm guessing that one can already be accessed directly in the action via vars.

File: .github/actions/run-e2e/action.yml [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+          echo "LC_ALL=C.UTF-8"
+          echo "IMAGE_REGISTRY=ghcr.io/hiero-ledger/hiero-mirror-node"
+          echo "BLOCK_NODE_VERSION=v0.35.1"
+          echo "CONSENSUS_NODE_VERSION=v0.75.1"
+          echo "HELM_RELEASE_NAME=mirror-1"
+          echo "SOLO_CLUSTER_NAME=test"
+          echo "SOLO_NAMESPACE=mirror"
+          echo "SOLO_CLUSTER_SETUP_NAMESPACE=solo"
+          echo "SOLO_DEPLOYMENT=solo-deployment"
+          echo "SOLO_VERSION=v0.77.0"
```
Comment (steven-sheehy) [inline .github/actions/run-e2e/action.yml:33]: Why can't we use `env` block to define these?

File: .github/actions/run-e2e/action.yml [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+              web3) chart_key=web3 ;;
+            esac
+            yq -i ".${chart_key}.image.registry = \"ghcr.io\"" mirror.yaml
+            yq -i ".${chart_key}.image.repository = \"hiero-ledger/hiero-mirror-node/${module}\"" mirror.yaml
+            yq -i ".${chart_key}.image.tag = \"${VERSION}\"" mirror.yaml
+            yq -i ".${chart_key}.image.pullPolicy = \"IfNotPresent\"" mirror.yaml
+          done
+          yq -i ".rest.image.tag = \"${TAG}\"" mirror.yaml
+          yq -i ".test.image.tag = \"${TAG}\"" mirror.yaml
+        fi
```
Comment (steven-sheehy) [inline .github/actions/run-e2e/action.yml:205]: In Solo v0.77.0, the images default to `ghcr.io/hiero-ledger/hiero-mirror-node/$module`. So I don't think this complicated logic is necessary. You can just delete the image tags: `yq -i 'del(.*.image)' mirror.yaml`. Then set `global.image.tag`.

File: .github/workflows/run-e2e.yml (hunk 13 above)
Comment (steven-sheehy) [reply .github/workflows/run-e2e.yml]: I think your comment doesn't apply to the latest version and you can now remove the ghcr login?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: .github/actions/run-e2e/action.yml (hunk 16 above)
Comment (bilyana-gospodinova) [reply .github/actions/run-e2e/action.yml]: `secrets: inherit` could be applied if run-e2e were a workflow. It is currently an action as requested in https://github.com/hiero-ledger/hiero-mirror-node/pull/13908#discussion_r3599373874. I removed these lines and added the login directly in the calling workflows.

File: monitor/src/main/java/org/hiero/mirror/monitor/publish/TransactionPublisher.java (hunk 15 above)
Comment (bilyana-gospodinova) [reply monitor/src/main/java/org/hiero/mirror/monitor/publish/TransactionPublisher.java]: Reverted.

File: .github/actions/run-e2e/action.yml (hunk 17 above)
Comment (bilyana-gospodinova) [reply .github/actions/run-e2e/action.yml:33]: In actions the `env` block can be set only per step, not globally for the whole action. I can move these only to the steps that use them but this way we will have a lot of duplication and it will be harder to track when we need to bump the solo version or the consensus node version, for example.

File: .github/workflows/acceptance.yaml (hunk 8 above)
Comment (bilyana-gospodinova) [reply .github/workflows/acceptance.yaml:32]: Reverted.

File: .github/workflows/run-e2e.yml (hunk 13 above)
Comment (bilyana-gospodinova) [reply .github/workflows/run-e2e.yml]: Removed

File: .github/actions/run-e2e/action.yml (hunk 18 above)
Comment (bilyana-gospodinova) [reply .github/actions/run-e2e/action.yml:205]: Addressed

**Review by steven-sheehy — APPROVED** [review]
